### PR TITLE
[24174] Filter work packages by date or range of date

### PR DESCRIPTION
**Pull request by silkentrance**
Filter by date and date range.

https://community.openproject.com/work_packages/24174

See also previous PR under #5071

### CODE REVIEW

**Comment by silkentrance** [comment]
@ulferts This includes the required tests, additional fixes and a little bit of refactoring.

@HDinger I managed to get the styles working as requested, however, in order to get this right, the whole grid layout of the filters needs to be redone. I am currently looking into foundation/grid model and see what I can come up with so that the filters become usable on phones and whatnot.

**Comment by silkentrance** [comment]
Cleaning up, fixing newly introduced codeclimate issues.

**Comment by silkentrance** [comment]
Rebased to current dev

**Comment by silkentrance** [comment]
fixing specs

trying to fix existing code climate issues with sql_for_field is rather difficult in that it always opens up a bunch of new issues. this needs to be refactored...

**Comment by silkentrance** [comment]
hell, there a lot of redundant travisci test runs...

**Comment by silkentrance** [comment]
@ulferts @lindenthal This now includes filtering by datetime / datetime ranges in the second commit. The tests for this are still missing, but you might want to have a look at it first and tell me your findings. I have used the RandomDataSeeder to create approx. 300 work packages with hourly advanced created_at dates and the filter seems to be working just fine across timezone boundaries.

In the UI, the displayed date(s) in the filter will be adjusted to the user's (configured) local timezone while on the server it will make sure that all users see exactly the same results. 

Existing codeclimate issues will be dealt with ASAP and if possible without me having to refactor the whole shebang.

Please note that I had to adjust bower.json to require newer versions of both moment and moment-timezone as there had been some issues with the older versions of the two packages.

**Comment by silkentrance** [comment]
@ulferts most code climate issues have been dealt with, the rest will require a more thorough refactoring and is beyond the scope of this PR.

**Comment by silkentrance** [comment]
@ulferts all tests are failing due to

```
/home/travis/build/opf/openproject/app/models/queries/relations.rb:32:in `<module:Relations>': uninitialized constant Queries::Queries::Register
```

Has this been fixed?

**Comment by silkentrance** [comment]
The following api endpoints have been removed

```
/api/experimental/projects/demo-project/work_packages?page=1&per_page=10
/api/experimental/projects/demo-project/queries/grouped
/api/experimental/projects/demo-project/queries/available_columns
```

This is the main reason that the tests are failing. I believe it is due to the overhaul of the query schema, so I will be patient :grinning:

**Comment by silkentrance** [comment]
@ulferts some unrelated tests still fail (relations) but the rest seems to be working just fine. perhaps you will try and have a go with internal testing for this branch?

**Review by ulferts — CHANGES_REQUESTED** [review]
@silkentrance I still propose splitting up filtering by date and filtering by datetime.

The PR does currently consist of three parts:
* Filtering by date
* Filtering by datetime
* styling fixes 

From what I have seen, filtering by date is working fine and we should be able to merge it soon. There are some minor things, e.g. there is no concatenating word like "and" between the two dates when filtering in an interval, but all those could also be addressed after the initial merge.

Filtering by datetime is IMO still not working where the source of error is still to be found on the conceptual level as Datetime fields are still filtered by a Date value, which leads to the strange behaviour of a user not being able to use the filter at all when he is located somewhere with noticeable time difference e.g. the USA, Brazil, China as filter results are split seemingly arbitrarily (because the user will not understand that he would have to change his time zone setting to UTC in order to have the filters and their results aligned.)
Below is a screenshot of such a situation:
image
I have stopped testing the feature but on top of the problem I see with the approach taken, reloading the page would lead to loosing the filter.

In order to concentrate on the Datetime problem, we should split up the PR and get the date feature along with the styling fixes into dev first. Then, the code study at each time for the Datetime feature would be way less. Because of this, I have not yet taken a deeper look into the code again.

**Comment by silkentrance** [comment]
@ulferts yes, the problem is communicating to the user that even though he enters a date, say, 2016-01-12 it is handled internally as a datetime object with timezone information, e.g. 2016-01-12T00:00:00-11:00. Similar for the range, however, here it will be two datetime objects, e.g. 2016-01-12T00:00:00-11:00 and 2016-01-12T23:59:59-11:00.

Now, what needs to be communicated to the user is that this will be converted to the server's local time zone and from these datetime object then the data will be filtered.

This also ensures that all users around the globe will get exactly the same result set.

For presentational purposes, datetime objects need to be displayed in the user's local timezone, then all datetime will be rendered correctly and you would no longer have the result from different dates problem.

As it is now, the implied UTC use case in the frontend, which I believe is a bad thing, will cause different dates to be shown.

If, on the other hand, the user configures a time zone in his or her settings, then the correct datetime will be shown.

So it is rather a question of why hold onto that implied UTC requirement when it clearly does not meet  real world requirements?

**Comment by ulferts** [comment]
> @silkentrance I still propose splitting up filtering by date and filtering by datetime.

Can we please do that and then talk about the Datetime implementation?

**Comment by silkentrance** [comment]
@ulferts This I can do, sure. However, I'd first like to see what I can do about the datetime impl, though. I think that I am determining the now() datetime object using just moment() instead of moment.utc(), which might lead to the observed problem. Let try to change that first, before going through the hustle of dissecting the individual changes.

**Comment by silkentrance** [comment]
@ulferts the changes regarding datetime (range) filtering have been removed.

**Comment by silkentrance** [comment]
@ulferts the changes for filtering datetime objects have been removed

**Comment by ulferts** [comment]
@silkentrance ah, now I understand that this PR does now only contain the Date functionality. I was under the impression that this would be done by #5155. I'll take a look, then.

File: frontend/app/components/filters/query-filters/query-filters.directive.html [hunk 1]
```diff
@@ -128,10 +156,11 @@
                                filter="filter"
                                is-multiselect="false"
                                isDisabled="isLoading"/>
-
         </div>
       </div>
 
+      </div>
```
Comment (ulferts) [inline frontend/app/components/filters/query-filters/query-filters.directive.html]: Didn't check the whole html tree but this `</div>` seems to be one to many.

File: app/models/queries/sql_for_field.rb [hunk 2]
```diff
@@ -78,43 +82,83 @@ def sql_for_field(field, operator, values, db_table, db_field, is_custom_filter
...
+    end
+    if to
+      to_date = Date.today + to
+    end
+    date_range_clause(table, field, from_date, to_date)
+  end
+
+  # Returns a SQL clause for date or datetime field for an exact range starting
+  # at the beginning of the day of from until the end of the day of to
   def date_range_clause(table, field, from, to)
```
Comment (ulferts) [inline app/models/queries/sql_for_field.rb:153]: We should cover this by tests

**Review by ulferts — CHANGES_REQUESTED** [review]
#### Unsupported operators for DateTime fields

The DateTime fields 'Updated on' and 'Created on' do still sport the 'on' and 'between' operators:

image

They should be removed for now until we are sure of how to support DateTime filtering correctly.

#### Console error

When selecting the 'on' or the 'between' operator, the following js error is logged:

image

There is probably required code in the removed commit.

**Comment by silkentrance** [comment]
Ups, nice find. These changes have been undone and the dangling closing div has been removed. As for the tests, I will implement these ASAP, but not this night :grin:

I also added the missing filterDateModelOptions, which had been removed during dissecting the previous commits.

**Comment by silkentrance** [comment]
@ulferts it seems that the existing tests do not cover such issues like for example missing operators and the frontend handling these missing operators, at all. But then again, let's postpone this to until the new query schema hits the road, which would make these tests much more simple and by that much more generic.

**Review by ulferts — APPROVED** [review]

**Comment by ulferts** [comment]
👍

**Pull request merged**